### PR TITLE
Replace default keyExtractor with individual key-fns

### LINKED ISSUE #3064: Explicitly define keyExtractor for list
[comment]: # (Please replace ... with your information. Remove < and >)

### User Story

As a developer, I want to be sure list elements don't behave funky because of incorrect `keyExtractor`.

### Description

*Type*: Bug

[list](https://github.com/status-im/status-react/blob/develop/src/status_im/ui/components/list/views.cljs) provides a default [keyExtractor](https://github.com/status-im/status-react/blob/develop/src/status_im/ui/components/list/views.cljs#L97) but it proved to be a source of [issues](https://github.com/status-im/status-react/pull/3061).

Remove it and make sure all `list` provide a correct `keyExtractor` as part of their definition (as `key-extractor`). FlatList [documentation](https://facebook.github.io/react-native/docs/flatlist.html) might provide some context.

### PR DESCRIPTION
**Pull request by voxlet**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

Fixes #3064 

### Summary:

- Remove default keyExtractor from `base-list-props`
- Add option to supply keyExtractor as `:key-fn`
- Wrap key-fn to coerce result to string
- Add individual key-fns to all `flat-list`s and `section-list`s

### Review notes (optional):

The change from `:key-extractor` to `:key-fn` is to allow function wrapping for string coercion. I will gladly remove this if preferred. 

I tried my best to supply an appropriate key-fn for all of the existing call sites, but where this was uncertain, I fell back to the item index.

### Steps to test:

- Verify the changed lists do not issue any warnings regarding children keys

status: ready

### CODE REVIEW

**Comment by jeluard** [comment]
@voxlet Great job! One suggestion: make all `key-fn` return string and modify the wrapper so that it does not coerce to string but validates using `:post` conditions. WDYT?

**Comment by voxlet** [comment]
@jeluard Thanks! WRT your suggestion, I can see that having the advantage of not redundantly coercing, while possibly having a negative effect on developer experience - figuring out what `:post` condition is being tripped, and having to litter `str` calls in `key-fn`s as a result. 

However, if we know the id-like values `key-fn` extracts are mostly strings to start with (which they are AFAICT), the downside would be negated accordingly. In this case, we would simply be looking to improve performance, so `spec/fdef` + dev build `spec/instrument` might be a better fit, so we can avoid the runtime hit in prod builds. (Do you already do instrumentation for dev builds?)

I wouldn't be against a `:post` condition to keep things simple, though. Probably more a matter of taste, hardly an issue that would warrant optimization in and of itself.

**Comment by jeluard** [comment]
@voxlet I think that's fine to enforce all `key-fn` to return strings. We could probably document that so it is explicit.

I understand `post` conditions are eluded in release mode?
Specs might be a better fit but we have't figured yet where we want to add specs and where not.
`post` might be the simplest option, at least as a first step.

**Comment by voxlet** [comment]
@jeluard 
> I understand post conditions are eluded in release mode?

[TIL!](https://clojure.org/reference/special_forms#__code_fn_code_name_em_param_em_em_condition_map_em_em_expr_em) Post conditions it is :)

This was a good decision, and surfaced some key-fns that were returning `nil` (subsequently coerced to `""`). Fixed these as well.

File: src/status_im/ui/screens/wallet/views.cljs [hunk 1]
```diff
@@ -66,7 +66,7 @@
      [list/flat-list
       {:default-separator? true
        :data               assets
-       :key-fn             :address
+       :key-fn             (comp str :symbol)
```
Comment (voxlet) [inline src/status_im/ui/screens/wallet/views.cljs]: `:address` returns `nil` for ETH

File: src/status_im/chat/screen.cljs [hunk 2]
```diff
@@ -79,7 +79,7 @@
   (letsubs [messages           [:get-current-chat-messages]
             current-public-key [:get-current-public-key]]
     [list/flat-list {:data                      messages
-                     :key-fn                    :message-id
+                     :key-fn                    #(or (:message-id %) (:value %))
```
Comment (voxlet) [inline src/status_im/chat/screen.cljs]: Date dividers don't have `:message-id`, fall back to `:value`

**Review by jeluard — APPROVED** [review]

**Comment by dmitryn** [comment]
>I understand post conditions are eluded in release mode?

@jeluard we should use https://cljs.github.io/api/compiler-options/elide-asserts compiler option for release, not sure if it's `true` by default

**Comment by jeluard** [comment]
@dmitryn Good point! 
@voxlet Do you mind adding the relevant config to the production configuration?

**Review by dmitryn — APPROVED** [review]

**Comment by dmitryn** [comment]
@jeluard what about making separate PR with `:elide-asserts true` compiler option?
For me it seems like out of scope of current PR.

**Comment by jeluard** [comment]
@dmitryn Makes sense!

**Comment by dmitryn** [comment]
@jeluard ok, i'll open PR soon.

**Review by alwx — APPROVED** [review]

**Comment by jeluard** [comment]
@voxlet Sorry you will have to rebase!

**Comment by voxlet** [comment]
@jeluard No problem!

**Comment by annadanchenko** [comment]
https://github.com/status-im/status-react/issues/3064#issuecomment-374302976
 @jeluard if you are fine with the implementation based on the review we can merge. Haven't found any functional bugs with "light weight" regression tests (send messages and ETH in 1:1 and group chats)

**Pull request merged**